File: DesignPatterns/6-mementoPattern/python/main.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-
###
# File: main.py
# Created Date: 2023-03-19 03:30:05
# Author: ysj
# Description:  备忘录模式客户端调用
###
from originator import RoleOriginator
from caretaker import CareTaker

# 主角
role = RoleOriginator(
    skills="大威天龙",
    hp=100,
    mp=100,
    attack=100,
    defense=100,
)

# 游戏状态托管人
caretaker = CareTaker()

# 保存进度
role.show_state("初始状态")
memento0 = role.create_memento()
caretaker.put(memento0)

# 攻打boss失败
role.hp = 0
role.mp = 10
role.attack = 1
role.defense = 20
role.show_state("攻打boss失败")

# 恢复之前的状态
memento0 = caretaker.take(0)
role.set_memento(memento0)
role.show_state("恢复初始状态")

# 打怪升级
role.skills = "超·大威天龙"
role.hp = 1000
role.mp = 1000
role.attack = 1000
role.defense = 1000

# 保存进度
role.show_state("打怪升级")
memento1 = role.create_memento()
caretaker.put(memento1)

# 挑战boss成功
role.hp = 2000
role.mp = 2000
role.attack = 2000
role.defense = 2000
role.show_state("挑战boss成功,属性翻倍")

# 保存进度
memento2 = role.create_memento()
caretaker.put(memento2)

# 想重新挑战boss
memento1 = caretaker.take(1)
role.set_memento(memento1)
role.show_state("重新挑战boss, 恢复之前状态")

File: DesignPatterns/6-mementoPattern/golang/main.go
/*
 * File: main.go
 * Created Date: 2023-03-19 02:31:23
 * Author: ysj
 * Description:  备忘录模式客户端调用
 */

package main

func main() {
	// 主角
	role := &RoleOriginator{
		Skills:  "大威天龙",
		HP:      100,
		MP:      100,
		Attack:  100,
		Defense: 100,
	}
	// 游戏状态托管人
	caretaker := &CareTaker{}

	// 保存进度
	role.ShowState("初始状态")
	memento0 := role.CreateMemento()
	caretaker.Put(memento0)

	// 攻打boss失败
	role.HP = 0
	role.MP = 10
	role.Attack = 1
	role.Defense = 20
	role.ShowState("攻打boss失败")

	// 恢复之前的状态
	memento0 = caretaker.Take(0)
	role.SetMemento(memento0)
	role.ShowState("恢复初始状态")

	// 打怪升级
	role.Skills = "超·大威天龙"
	role.HP = 1000
	role.MP = 1000
	role.Attack = 1000
	role.Defense = 1000

	// 保存进度
	role.ShowState("打怪升级")
	memento1 := role.CreateMemento()
	caretaker.Put(memento1)

	// 挑战boss成功
	role.HP = 2000
	role.MP = 2000
	role.Attack = 2000
	role.Defense = 2000
	role.ShowState("挑战boss成功,属性翻倍")

	// 保存进度
	memento2 := role.CreateMemento()
	caretaker.Put(memento2)

	// 想重新挑战boss
	memento1 = caretaker.Take(1)
	role.SetMemento(memento1)
	role.ShowState("重新挑战boss, 恢复之前状态")
}
